crates/ruff_python_parser/src/token_source.rs

```diff
@@ -139,13 +139,17 @@ impl<'src> TokenSource<'src> {
 
     /// Consumes the token source, returning the collected tokens and any errors encountered during
     /// lexing. The token collection includes both the trivia and non-trivia tokens.
-    pub(crate) fn finish(self) -> (Vec<Token>, CommentRanges, Vec<LexicalError>) {
+    pub(crate) fn finish(mut self) -> (Vec<Token>, CommentRanges, Vec<LexicalError>) {
         assert_eq!(
             self.current_kind(),
             TokenKind::EndOfFile,
             "TokenSource was not fully consumed"
         );
 
+        if let Some(last) = self.tokens.pop() {
+            assert_eq!(last.kind(), TokenKind::EndOfFile);
+        }
+
```

Comment:
The `EndOfFile` token shouldn't be included in the token stream, it's just to indicate the parser to stop. This isn't in `bump` because it only needs to be done once.

Reply:
I think this comment would be useful in the source. I personally wouldn't mind if the EOF token is in the token stream but I don't feel strongly (and I wouldn't make the change as part of this stack if it changes too many snapshot tests)